mixnet/mixnet.py

```diff
@@ -12,31 +16,105 @@
 class Mixnet:
     mix_nodes: List[MixNode]
 
-    # Build a new topology deterministically using an entropy.
-    # The entropy is expected to be injected from outside.
-    #
-    # TODO: Implement constructing a new topology in advance to minimize the topology transition time.
-    #       https://www.notion.so/Mixnet-Specification-807b624444a54a4b88afa1cc80e100c2?pvs=4#9a7f6089e210454bb11fe1c10fceff68
-    def build_topology(
+    def __init__(
         self,
-        entropy: bytes,
+        mix_nodes: List[MixNode],
         n_layers: int,
         n_nodes_per_layer: int,
+        mixnet_epoch_sec: int,
+        proactive_sec: int,
+        initial_entropy: bytes,
+    ):
+        self.mix_nodes = mix_nodes
+
+        self.set_entropy(initial_entropy)
+
+        self.topology_updater = MixnetTopologyUpdater(
+            mix_nodes, n_layers, n_nodes_per_layer, mixnet_epoch_sec, proactive_sec
+        )
+        self.topology_updater.daemon = True
+        self.topology_updater.start()
+
+    def set_entropy(self, entropy: bytes) -> None:
+        """
+        Set entropy (seed) to the Fisher-Yates shuffler
+        that is going to be used for deterministic topology construction.
+
+        This method is expected to be called periodically by the user.
+        Until the next entropy is provides, Fisher-Yates shuffler continues to works
+        based on this entropy set once.
+        """
+        FisherYates.set_seed(entropy)
+
+    def choose_mixnode(self) -> MixNode:
+        return random.choice(self.mix_nodes)
+
+    def current_topology(self) -> MixnetTopology:
+        return self.topology_updater.current_topology()
+
+
+class MixnetTopologyUpdater(Thread):
+    def __init__(
+        self,
+        mix_nodes: List[MixNode],
+        n_layers: int,
+        n_nodes_per_layer: int,
+        mixnet_epoch_sec: int,
+        proactive_sec: int,
+    ):
+        super().__init__()
+        self.mix_nodes = mix_nodes
+        self.n_layers = n_layers
+        self.n_nodes_per_layer = n_nodes_per_layer
+        self.mixnet_epoch_sec = mixnet_epoch_sec
+        self.proactive_sec = proactive_sec
+
+        self.lock = threading.Lock()
+        self.topology = self.build_topology()
+        self.next_topology = None
+
+    def run(self) -> None:
+        next_epoch_ts = datetime.now() + timedelta(seconds=self.mixnet_epoch_sec)
```

Comment:
Thinking at loud here.

I don't know if this is the best way to count epochs, it may work here when all node instances have their clocks synchronized. However, when one node will connect later (or will disconnect for some time) he might loose track of time here. It might be better to have something like `self.genesis_timestamp` from which all nodes calculated current epoch:
`epoch_number = ceiling(datetime.now() - params.genesis_timestamp / params.epoch_len_in_sec)`
and slot number as (relative to the epoch):
`slot_number_in_epoch = ceiling((datetime.now() - epoch_number * params.epoch_len_in_sec) / params.slot_len_in_sec)`
Or something similar.

This way the calculation should be independent from the connectivity conditions of nodes and can be influenced only by clock drifts.

Another things is that the epoch and slot information should be confirmed by reading them from the consensus. But that might be out of the scope for this specification effort.

Reply:
Yes. I missed this point. I'll fix it.

> Another things is that the epoch and slot information should be confirmed by reading them from the consensus. But that might be out of the scope for this specification effort.

I'm also thinking about this. Here, we only use `mixnet_epoch_sec` that defines the interval of topology updates, which is going to be shorter than `consensus_epoch_sec`. So, if we design the entropy to be injected from the consensus layer to the mixnet layer, I think the mixnet layer doesn't need to read something (consensus epoch or slot info) from the consensus layer. Did I understand the point that you mentioned?

Reply:
Yes, I think that we are thinking the same way here. I've left more in depth comment below in the next conversation (https://github.com/logos-co/nomos-specs/pull/51#discussion_r1459185382)